nil-discussion.md

```diff
@@ -0,0 +1,180 @@
+# nil values and validation
+
+`nil` setters and property defaults are a big snarl with our system because of
+this conflict:
+
+1. People who don't specify a default value for their property expect the
+   default to be `nil`.
+2. Some people want `nil` to be an invalid value for their property so that they
+   can simplify their code by not having to handle it.
+3. Copying a property from one resource to another is a common thing, and the
+   combination of the above rules currently causes `property :x, String` to
+   have an invalid default value, meaning it can't be copied. This is bad.
+
+In coming up with solutions, there are a number of things around default values
+and validation that have to be considered. I've written down these things (the
+things I think we want to be true) here, and enumerated all of the solutions
+I've heard so far, along with which rules they violate and surprises they
+create.
+
+## Rules
+
+Things we want to be true:
+
+1. A property with no default value successfully returns `nil` as its default,
+   regardless of validation.
+
+   This is the crux of the issue, the exception that means we have to do
+   *something* special.
+
+   - `property :x, String` has `nil` as its default value.
+   - `property :x` has `nil` as its default value.
+
+2. The default value of a property must be settable (the copy rule).
+
+   Corollary: Setters and default values are validated, so nothing can get *into*
+   the property unless it is valid.
+
+   - `x("10")` for `property :x, Integer` is INVALID.
+   - `property :x, Integer, default: "10"` is INVALID.
+   - `property :x, Integer, default: lazy { "10" }` is INVALID.
```

Comment:
I would accept a system where this isn't checked but I agree it should be invalid at least in the abstract.